content/2019-07-04_vyvijime-rozhrani.md

```diff
@@ -0,0 +1,50 @@
+Jedná se o čas, který potřebuje server na zpracování karet. Může se stát, že serveru dojde interní časovač, ale ještě než dá o tom vědět klientům, dostane další zprávu s aktualizovanými kartami. Protože časové okénko na výběr už je zavřené, karty se přiřadí až k dalšímu kolu. Abychom tomu zamezily, je dobré přidat do zpráv s kartami číslo herního kola. Tím si zajistíme korektní přiřazení všech zpráv.
+
+## Kolik stavů potřebujeme? A co patří k Interface?
+
+Nyní vytváříme stav hry a stav rozhraní. Stav hry drží informace o mapě a robotech, zatímco stav rozhraní patří konkrétnímu hráči a ukazuje detaily jeho robota + mini obrázky ostatních hráčů. 
+Řešily jsme, zda klientovi, který se zabývá vykreslením rozhraní, máme posílat i celkový stav hry. Shodly jsme se, že ano, jelikož data o všech robotech musíme "vyzobat" a aktualizovat ze stavu hry.
+
+U stavu rozhraní máme další zajímavou věc: obsah jeho atributů totiž proudí mezi serverem a klientem: informaci o zvolených kartách, stavu `power_down` a ukončení výběru posílá hráč serveru. Data o robotovi a karty, ze kterých si může hráč vybrat, posílá hráči server.
+Je důležité proto odesílané zprávy pečlivě naplánovat a vždy důkladně promýšlet, která data posíláme a která přijímáme.
+
+Když se podíváme na stávající rozhraní, vykreslujeme ho hned s příkladovým (_fake_) robotem a kartami. 
+Může se ale stát, že chceme vykreslit interface, ale ještě neznáme svého robota ani nemáme karty. Interface se proto musí umět vykreslit i v případě, že tyto údaje nejsou k dispozici.
+
+Co víc, máme v souboru s rozhraním i nějaké zavolané funkce: vytváří se v něm balíček karet, karty se rozdají a vykreslí, a vytvoří i instance třídy samotného interface. Tyto exekuce v souboru s logikou chování nechceme, dělá to kód netestovatelný a začíná se nám to přít s kódem, který postupně připisujeme serveru. Logika vytváření karet, které nepatří k logice klienta, taky nepatří tomuto souboru, měla by se přenést pod stav hry.
+
+## Klient Interface
+
+Jelikož náš klient obsahuje v sobě už dost kódu a začaly jsme využívat globální proměnné, aby vše fungovalo, jak si představujeme, bylo nám doporučeno překlopit stávající proceduru na třídu. Globální proměnné tak překlopíme do atributů třídy, se kterými pak můžeme jednodušeji pracovat. U takového "roztřídění" kódu je velice podstatné rozmyslet si, které atributy patří nové vytvořené třídě (`self.window`, `self.state`) a které jsou dané zvenčí (v našem případě: `text` ve smyslu vstupu z klávesnice). Pak zbývá rozchodit volání metod (spousta zapomenutých `self` po cestě, které mají za výsledek `NameError`).
+
+## Kam patří robotí program?
+
+Petr sugeroval, že seznam karet, které má robot na ruce, možná vůbec robotovi nepatří. Provedení efektů karet a políček (funkce `apply_all_effects`) by měla dostat jako argumenty roboty a říct, co s nimi má udělat. `Program` má totiž smysl tehdy, když se provádí, samotný není vlastnost robota. 
```

Comment:
Nejsem si úplně jistá, co zde myslíš slovem "sugeroval" - v češtině to má význam "vnutit někomu nějakou myšlenku", něco někomu namlouvat. Je to to, co jsi myslela? Nebo spíš, že to navrhoval?

Reply:
zmenila jsem na "podotkl", ktere asi nejlip vystihuje to, co jsem chtela vyjadrit - trochu mene nez navrh, proste sugestie. ;)